### PR TITLE
[None][feat] Visual Gen: add cuda graphs; torch compile; nvtx; warmup

### PR DESCRIPTION
**Pull request by NVShreyas**
## Summary by CodeRabbit

* **New Features**
  * Added TorchCompile support with multiple compilation modes (default, max-autotune, reduce-overhead) and fullgraph option for flexible optimization
  * Introduced CUDA graph optimization to improve inference performance
  * Added configurable warmup steps for model initialization and compilation
  * Integrated layer-wise NVTX markers for detailed performance profiling and analysis

## Description
Performance:

WAN 2.1 w/ torch compile + warmup: 30-90% lower latency
WAN 2.2 w/torch compile + warmup: 15-60% lower latency

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/config.py [hunk 1]
```diff
@@ -196,15 +196,20 @@ class PipelineConfig(BaseModel):
     """General pipeline configuration."""
 
     enable_torch_compile: bool = True
-    torch_compile_models: str = PipelineComponent.TRANSFORMER
-    torch_compile_mode: str = "default"
+    torch_compile_models: List[str] = [PipelineComponent.TRANSFORMER]
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/config.py]: How about models with multiple transformer blocks, for example Wan's two-transformer model?

File: examples/visual_gen/visual_gen_wan_i2v.py [hunk 2]
```diff
@@ -161,6 +196,15 @@ def main():
             "dit_cfg_size": args.cfg_size,
             "dit_ulysses_size": args.ulysses_size,
         },
+        "pipeline": {
+            "enable_cuda_graph": args.enable_cudagraph,
+            "enable_torch_compile": not args.disable_torch_compile,
+            "torch_compile_models": args.torch_compile_models,
+            "torch_compile_mode": args.torch_compile_mode,
+            "enable_fullgraph": args.enable_fullgraph,
```
[the remark this answers is not in the document]
Comment (NVShreyas) [reply examples/visual_gen/visual_gen_wan_i2v.py]: added

**Review by chang-l — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/visual_gen/models/wan/pipeline_wan.py [hunk 3]
```diff
@@ -241,6 +242,33 @@ def post_load_weights(self) -> None:
...
+        Runs warmup inference with common shapes for Wan models.
+        """
+        common_shapes = [(480, 832, 33), (720, 1280, 81)]
+        
+        if self.is_wan22:
+            # Double warmup steps to also warmup the 2nd transformer
+            warmup_steps = warmup_steps * 2
+
+        for height, width, num_frames in common_shapes:
+            logger.info(f"Warmup: Wan {height}x{width}, {num_frames} frames {warmup_steps} steps")
```
Comment (NVShreyas) [inline tensorrt_llm/_torch/visual_gen/models/wan/pipeline_wan.py]: @chang-l I'd like to know your thoughts here. Its not easy to know which model/resolution is going to be served so I run warmup over all of them. I think for serving scenario this should be the right thing as then user can request images with different H, W and F. But for single offline request, its not ideal and we take 0.5s latency hit but I don't think there's a good way to get around it for now.
This also increases the server initialization time but I believe that shouldn't matter too much

Comment (chang-l) [reply tensorrt_llm/_torch/visual_gen/models/wan/pipeline_wan.py]: Thanks for bringing this up. Yes, adding common shapes to the warm-up to offset compilation cost from the user's real inference time makes sense to me.
Maybe we could also have some warning if the user passes T/H/W values that differ from the warm-up shapes, to remind them they may observe a longer first denoising step due to torch.compile recompilation

**Review by o-stoner — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]

File: tests/unittest/_torch/visual_gen/test_model_loader.py [hunk 4]
```diff
@@ -434,6 +440,7 @@ def test_fp8_vs_bf16_memory_comparison(checkpoint_exists):
         checkpoint_path=CHECKPOINT_PATH,
         quant_config={"quant_algo": "FP8_BLOCK_SCALES", "dynamic": True},
         skip_components=SKIP_HEAVY_COMPONENTS,
+        pipeline={"warmup_steps": 0},
```
Comment (zhenhuaw-me) [inline tests/unittest/_torch/visual_gen/test_model_loader.py]: Does it make sense to keep 1 test with the "default (or mostly used) warmup steps" to protect the general use case? (also apply to other tests)

File: tensorrt_llm/_torch/visual_gen/models/wan/pipeline_wan.py [hunk 5]
```diff
@@ -241,6 +247,32 @@ def post_load_weights(self) -> None:
...
+        for height, width, num_frames in self.common_warmup_shapes:
+            logger.info(f"Warmup: Wan {height}x{width}, {num_frames} frames {warmup_steps} steps")
+
+            with torch.no_grad():
+                self.forward(
+                    prompt="warmup",
+                    negative_prompt="",
+                    height=height,
+                    width=width,
+                    num_frames=num_frames,
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/wan/pipeline_wan.py]: https://jirasw.nvidia.com/browse/TRTLLM-11112 to track the improvements

File: tensorrt_llm/_torch/visual_gen/executor.py [hunk 6]
```diff
@@ -186,6 +187,16 @@ def serve_forever(self):
 
     def process_request(self, req: DiffusionRequest):
         """Process a single request."""
+        if (
+            self.pipeline.common_warmup_shapes
+            and (req.height, req.width, req.num_frames) not in self.pipeline.common_warmup_shapes
+        ):
+            logger.warning(
+                f"Requested shape (height={req.height}, width={req.width}, num_frames={req.num_frames}) "
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/executor.py]: The fix will be https://jirasw.nvidia.com/browse/TRTLLM-11104

File: examples/visual_gen/visual_gen_wan_i2v.py [hunk 7]
```diff
@@ -126,6 +126,46 @@ def parse_args():
...
+    parser.add_argument(
+        "--enable_cudagraph", action="store_true", help="Enable CudaGraph acceleration"
+    )
+
+    # torch compile
+    parser.add_argument(
+        "--disable_torch_compile", action="store_true", help="Disable TorchCompile acceleration"
+    )
+    parser.add_argument(
+        "--torch_compile_models",
```
Comment (zhenhuaw-me) [inline examples/visual_gen/visual_gen_wan_i2v.py]: [TRTLLM-11116](https://jirasw.nvidia.com/browse/TRTLLM-11116) to follow up

**Review by zhenhuaw-me — COMMENTED** [review]
Thanks for creating this foundation support in the `main` branch. I filed several JIRAs to follow up for the sake of user experience.
